### ISSUE TITLE
PyTorch Distributed Training

### ISSUE THREAD

**Issue by ZhuoranYu** [kekmodel/FixMatch-pytorch#16]
Hello,

Thanks for your good work! I really enjoyed your implementation. While I was trying to reproduce your results, I found some questions about PyTorch distributed training:
1. Why don't you reduce losses across GPUs before calling the backward? I think typically you may want to call torch.distributed.all_reduce to synchronize the gradients across the GPUs. Is it just missing from the code or you have some equivalent operations that I have overlooked?
2. When I run the code on single GPU and multiple GPUs, the distributed training does not make the training process faster(even slower). The overall batch sizes are consistent in both settings(64 for single GPU and 4*16 for 4 GPUs). Is this expected?

Thanks for your help in advance!

**Comment by zijian-hu** [comment]
Regarding

> 1. Why don't you reduce losses across GPU ...

You don't need to reduce loss manually. The loss is reduced automatically when you call `loss.backward()`.

As described in [this answer](https://discuss.pytorch.org/t/distributeddataparallel-loss-compute-and-backpropogation/47205/2?u=ciketop)

> on loss.backward() all processes reduce their gradients. As loss.backward() returns, the gradients of your model parameters will be the same, and the optimizer in each process will perform the exact same update to the model parameters

<br>
Regarding

> 2. When I run the code on single GPU and multiple GPUs ...

Distributed training does not always make training faster. In my opinion, the point of using distributed training is to have a larger effective batch size that is otherwise too large for a single GPU. It is normal that when you have a small batch size (i.e. 16 per GPU in your case), distributed is slower than a single GPU since there are communication overheads between different processes (assume you use 1 process per GPU).

Also, distributed training could degrade performance. See [this answer](https://discuss.pytorch.org/t/training-performance-degrades-with-distributeddataparallel/47152/16) for detail. You might need to use a different learning rate.
